### PR TITLE
implementing carmin data

### PR DESCRIPTION
**Pull request by alaeessaki**
This PR contains carmin data download endpoint implementation.

NB : the utility class is a POJO. changing it to component caused problems with the NIFTI download and it complicated the unit testing.

### CODE REVIEW

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/dataset/controler/CarminDataApi.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+ */
+@Api(value = "carmin-data")
+@RequestMapping("/carmin-data")
+public interface CarminDataApi {
+
+    @ApiOperation(value = "Get content or information for a given path", notes = "Download a file (or a directory) or retun information about a specific path. The response format and content depends on the mandatory action query parameter (see the parameter description). Basically, the \"content\" action downloads the raw file, and the other actions return various information in a JSON record.", tags={  })
+    @ApiResponses(value = {
+            @ApiResponse(code = 200, message = "successful response. If the action is \"content\", the raw file (or a tarball) is returned, with the according mime type. Otherwise a json response a returned", response = GetPathResponse.class),
+
+            @ApiResponse(code = 200, message = "A functional or internal error occured processing the request") })
```
Comment (michaelkain) [inline shanoir-ng-datasets/src/main/java/org/shanoir/ng/dataset/controler/CarminDataApi.java:42]: Hi Alae, I hope you are doing fine. Could you please adapt the return code of the error to 500? Thank you and with kind regards, Michael

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/dataset/controler/CarminDataApi.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+ * @author Alae Es-saki
+ */
+@Api(value = "carmin-data")
+@RequestMapping("/carmin-data")
+public interface CarminDataApi {
+
+    @ApiOperation(value = "Get content or information for a given path", notes = "Download a file (or a directory) or retun information about a specific path. The response format and content depends on the mandatory action query parameter (see the parameter description). Basically, the \"content\" action downloads the raw file, and the other actions return various information in a JSON record.", tags={  })
+    @ApiResponses(value = {
+            @ApiResponse(code = 200, message = "Successful response. If the action is \"content\", the raw file (or a tarball) is returned, with the according mime type. Otherwise a json response a returned", response = GetPathResponse.class),
+            @ApiResponse(code = 200, message = "A functional or internal error occured processing the request"),
```
Comment (michaelkain) [inline shanoir-ng-datasets/src/main/java/org/shanoir/ng/dataset/controler/CarminDataApi.java:42]: Dear Alae, I hope you are doing fine. Two lines with status code 200 does not seem to be correct. Could you please adapt the change? Thank you and with kind regards, Michael

Comment (alaeessaki) [reply shanoir-ng-datasets/src/main/java/org/shanoir/ng/dataset/controler/CarminDataApi.java]: hello Michael, I hope you are doing well, 
Can you check the new changes?
Thanks.

Comment (alaeessaki) [reply shanoir-ng-datasets/src/main/java/org/shanoir/ng/dataset/controler/CarminDataApi.java]: Hello michael, do you have any idea about this error in github pipes ? what ive did, is only deleting one line of the apiResponse. and now the pipe can't build the application with an error in a line that doesn't exist in the file.

image

image

Comment (alaeessaki) [reply shanoir-ng-datasets/src/main/java/org/shanoir/ng/dataset/controler/CarminDataApi.java]: when i run the build on my machine it works fine 

image

Comment (alaeessaki) [reply shanoir-ng-datasets/src/main/java/org/shanoir/ng/dataset/controler/CarminDataApi.java:42]: @michaelkain , i rolledback to the last working pipe commit, and the error is still showing.
what do you think ?

Comment (alaeessaki) [reply shanoir-ng-datasets/src/main/java/org/shanoir/ng/dataset/controler/CarminDataApi.java]: hello @michaelkain , i've found that in the dev branch, there is a new calling for a method moved to the utility file in this PR, so the compiler doesn't recognise it
I pulled the dev and it's fixed.

**Comment by michaelkain** [comment]
Dear Alae, thank you for this PR.
It is fine now for the status codes. Unfortunately there might be something weird with your PR.
1) When I try to pull it to my local develop branch, that is based on develop fli-iam, I can not do the merge with fast-forwarding, as I have an error. I have to de-activated fast-forward to force the merge.
2) When I bootstrap.sh --clean with your PR, I have only one user called "admin" created in Keycloak, what is very strange.
None of the 3 typical users is created in the database, but on GitHub it does not show any change on any related file...
Do you have an explanation for that?
With kind regards, Michael

**Comment by alaeessaki** [comment]
hello @michaelkain,
thank you for your response, my response to the two points :

1- I pulled the new dev branch and fixed the merge conflict before the last push of this PR. 
ps: I tried to recreate the error, the merge from this branch to the develop branch doesn't log any abnormal behavior. (the development branch in my local is up to date with the origin repos).
2- This is the same error that @KhalilKes faced and reported by mail ; 
start of mail : 
```
Dear Michael,

First I apologize for the late PR, but I am actually facing an issue while running the bootstrap.sh, either with the force flag or the clean one, it stops at the Keycloak container creation which doesn't create the users, I don't know if you ever encountered this behavior before.

Looking forward for your kind reply
```
end of mail.

ps: he tried to run bootstrap.sh with clean and force, but it doesn't create any user.
in my opinion, I don't think it's related to this PR, since it only implements Carmin dataset download.

With kind regards, Alae

**Comment by michaelkain** [comment]
Dear Jean-Come (@jcomedouteau),
could you please test this PR on your local machine? For me, it was very strange to have in Keycloak only one user, assigned to the realm, instead of having 3 users. Could you please verify, if it is the same on your machine? Thank you and with kind regards, Michael

**Comment by michaelkain** [comment]
Dear Alae, thank you for your answer and sorry for coming back late on this. I was not sure, what to do next.
I would like to have the opinion of Jean-Come please. Having an error or no user in Keycloak I could understand, but having only one user, called admin, in Keycloak, attached to the realm, and none of your files touches to that area, is strange. Especially, when I force your PR and build it, I have that behaviour, when I go back to develop and compile, the behaviour is gone. And there is the error, that I can not fast-forward your PR as well. Cheers and we keep in touch, Michael

**Comment by alaeessaki** [comment]
Hello @michaelkain,
Thank you for your comment.
Can I know where is the file that touches that area? 
For the fast forward problem, I think it's related to my rebase of one of my commits. my suggestion to fix this problem is that I can create a new branch from develop, and I'll do those changes and create a new PR. with this, we'll be sure that it's not related to the fast-forward.

With kind regards, Alae

**Comment by michaelkain** [comment]
Dear Alae, the file concerned is https://github.com/fli-iam/shanoir-ng/blob/develop/shanoir-ng-users/src/main/resources/scripts/import.sql. It might be a brilliant idea to start from develop and just pick your modifications. Cheers and thank you a lot, Michael

**Comment by jcomedouteau** [comment]
Hi @michaelkain 

The 3 default users are well created in my keycloak:
image

Maybe you were in the wrong realm ? -> You have only one admin user in the "master" realm of keycloak and not in shanoir realm:
image

**Comment by michaelkain** [comment]
Hi Alae, hi Jean-Come,
@jcomedouteau was right, now I have 3 users in Keycloak in realm shanoir.
Will merge it and sorry for the huge delay! With kind regards, Michael

**Review by michaelkain — APPROVED** [review]

**Pull request merged**